### PR TITLE
Develop update

### PR DESCRIPTION
**Pull request by maksud**
* Updated emulators
* Checked accuracy and reproducible results 
* Added `symamd` function from Suite Sparse
* Updated sort algorithms (insertion and quick sort)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,7 @@ set(PACKAGE_NAME  "HyKKT")
 set(PACKAGE_TARNAME "hykkt")
 
 option(HYKKT_TEST_WITH_BSUB "Use `jsrun` instead of `mpirun` commands when running tests" OFF)
+option(HYKKT_USE_AMD "Use AMD from SuiteSparse" OFF)
```
Comment (pelesh) [inline CMakeLists.txt:12]:
```suggestion
option(HYKKT_USE_AMD_ORDERING "Use AMD from SuiteSparse" OFF)
```
Specify that "AMD" stands for approximate minimum degree ordering; not for AMD GPUs.

Comment (cameronrutherford) [reply CMakeLists.txt:12]: Can HyKKT eventually be used with AMD GPUs?

File: CMakeLists.txt [hunk 2]
```diff
@@ -28,6 +29,22 @@ include(CheckLanguage)
 enable_language(CUDA)
 check_language(CUDA)
 
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
+configure_file(
+  "${CMAKE_SOURCE_DIR}/src/hykkt_defs.hpp.in"
+  "${CMAKE_BINARY_DIR}/hykkt_defs.hpp")
+include_directories(${CMAKE_BINARY_DIR})
```
Comment (cameronrutherford) [inline CMakeLists.txt:36]: I think we should avoid blanket include_directories, but I don't think we follow that in ExaGO or HiOp, so no big deal

File: cmake/FindAMD.cmake [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  target_link_libraries(AMD INTERFACE ${AMD_LIBRARY})
+  target_include_directories(AMD INTERFACE ${AMD_INCLUDE_DIR})
+  get_filename_component(AMD_LIB_DIR ${AMD_LIBRARY} DIRECTORY)
+  set(CMAKE_INSTALL_RPATH "${AMD_LIB_DIR}")
+else()
+  message(STATUS "AMD was not found.")
+endif()
+
+set(AMD_INCLUDE_DIR CACHE PATH "Path to amd.h")
+set(AMD_LIBRARY CACHE PATH "Path to amd library")
```
Comment (cameronrutherford) [inline cmake/FindAMD.cmake:38]: We should consider adding Incline testing to CI if this builds on ROCm platforms

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -133,4 +135,22 @@ install(TARGETS chol_driver DESTINATION "${CMAKE_INSTALL_PREFIX}")
 install(TARGETS schur_cg_driver DESTINATION "${CMAKE_INSTALL_PREFIX}")
 install(TARGETS hybrid_driver DESTINATION "${CMAKE_INSTALL_PREFIX}")
 install(TARGETS solver_driver DESTINATION "${CMAKE_INSTALL_PREFIX}")
-
+install(TARGETS solver_driver_emulator DESTINATION "${CMAKE_INSTALL_PREFIX}")
```
Comment (cameronrutherford) [inline examples/CMakeLists.txt:138]: We should probably add this driver as a test to CMake as well

File: examples/CMakeLists.txt [hunk 5]
```diff
@@ -133,4 +135,22 @@ install(TARGETS chol_driver DESTINATION "${CMAKE_INSTALL_PREFIX}")
...
+
+
+if (HYKKT_USE_AMD)
+  if(NOT AMD_LIBRARY)
+    message(STATUS "Cannot find AMD, disabling SuiteSparse AMD module ...")
+    set(HYKKT_USE_AMD OFF CACHE BOOL "Build without SuiteSparse AMD module." FORCE)
+  else()
+    target_link_libraries(perm_driver PUBLIC AMD)
```
Comment (cameronrutherford) [inline examples/CMakeLists.txt:146]: We should either use a list here, or just have the AMD library be linked in through the HyKKT library itself

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -39,3 +39,14 @@ install(TARGETS hykkt_libs EXPORT HyKKT-targets)
 add_library(HyKKT::HyKKT ALIAS hykkt_libs)
 
 
+if (HYKKT_USE_AMD)
+  if(NOT AMD_LIBRARY)
+    message(STATUS "Cannot find AMD, disabling SuiteSparse AMD module ...")
+    set(HYKKT_USE_AMD OFF CACHE BOOL "Build without SuiteSparse AMD module." FORCE)
+  else()
+    target_link_libraries(hykkt_libs PUBLIC AMD)
```
Comment (cameronrutherford) [inline src/CMakeLists.txt:47]: HyKKT libs already links agains the AMD library, so the examples don't also need to link the lib

File: src/permcheck.cu [hunk 7]
```diff
@@ -62,7 +114,12 @@ void make_vec_map_c(int n,
       perm_map[row_s + j]  = row_s + j;
       perm_cols[row_s + j] = rev_perm[cols[row_s + j]];
     }
+#if 0
     selection_sort2(rowlen, &perm_cols[row_s], &perm_map[row_s]);
+#else
+    //quickSort(&perm_cols[row_s], &perm_map[row_s], 0, rowlen-1);
+    insertion_sort(rowlen, &perm_cols[row_s], &perm_map[row_s]);
```
Comment (cameronrutherford) [inline src/permcheck.cu:121]: I have a suggestion for an alternative sort. Nothing to change here, but I think I should create an issue for it once I'm back at my computer...

**Review by cameronrutherford — CHANGES_REQUESTED** [review]
Few small changes to CMake, and an issue I need to create

File: src/CMakeLists.txt (hunk 6 above)
Comment (pelesh) [reply src/CMakeLists.txt:47]: AMD stands here for Approximate Minimum Degree ordering library from SuiteSparse.

File: CMakeLists.txt (hunk 1 above)
Comment (pelesh) [reply CMakeLists.txt:12]: HyKKT does not build on AMD GPU platforms as yet, in this context AMD is Approximate Minimum Degree library from SuiteSparse.

File: src/permcheck.cu (hunk 7 above)
Comment (pelesh) [reply src/permcheck.cu:121]: This is probably outside the scope of this PR, so I will resolve this conversation, but I look forward to the issue on alternative sort options.

File: cmake/FindAMD.cmake (hunk 3 above)
Comment (pelesh) [reply cmake/FindAMD.cmake:38]: HyKKT has not been ported to HIP/ROCm yet.

**Review by cameronrutherford — APPROVED** [review]

**Pull request merged**